File: packages/surface-filters/src/pixelate.ts
import type { SurfaceRegion } from '@flighthq/types';

/**
 * Pixelates (mosaics) `source` into `out`: the region is divided into
 * `blockSize × blockSize` cells, each cell is averaged, and every output pixel
 * in the cell takes that average color. `blockSize` is clamped to at least 1
 * (a no-op copy at 1). Edge cells are clipped to the region.
 *
 * `out` must be at least `source.width * source.height * 4` bytes. Safe to pass
 * `source.surface.data` as `out` for a full-surface region — each cell is fully
 * read before it is written, and cells do not overlap.
 */
export function applySurfacePixelateFilter(
  out: Uint8ClampedArray,
  source: Readonly<SurfaceRegion>,
  blockSize: number,
): void {
  const block = Math.max(1, Math.round(blockSize));
  const w = source.width;
  const h = source.height;
  const surfaceWidth = source.surface.width;
  const surfaceHeight = source.surface.height;
  const data = source.surface.data;

  for (let by = 0; by < h; by += block) {
    const yEnd = Math.min(by + block, h);
    for (let bx = 0; bx < w; bx += block) {
      const xEnd = Math.min(bx + block, w);
      let r = 0;
      let g = 0;
      let b = 0;
      let a = 0;
      let count = 0;
      for (let py = by; py < yEnd; py++) {
        const sy = source.y + py;
        if (sy < 0 || sy >= surfaceHeight) continue;
        for (let px = bx; px < xEnd; px++) {
          const sx = source.x + px;
          if (sx < 0 || sx >= surfaceWidth) continue;
          const si = (sy * surfaceWidth + sx) * 4;
          r += data[si];
          g += data[si + 1];
          b += data[si + 2];
          a += data[si + 3];
          count++;
        }
      }
      if (count === 0) continue;
      const ar = Math.round(r / count);
      const ag = Math.round(g / count);
      const ab = Math.round(b / count);
      const aa = Math.round(a / count);
      for (let py = by; py < yEnd; py++) {
        for (let px = bx; px < xEnd; px++) {
          const di = (py * w + px) * 4;
          out[di] = ar;
          out[di + 1] = ag;
          out[di + 2] = ab;
          out[di + 3] = aa;
        }
      }
    }
  }
}

File: crates/flighthq-surface-filters/src/pixelate.rs
//! Pixelate (mosaic) filter — block-averaging to produce a pixelated effect.

use flighthq_types::SurfaceRegion;

/// Pixelates (mosaics) `source` into `out`: the region is divided into
/// `block_size × block_size` cells, each cell is averaged, and every output
/// pixel in the cell takes that average color. `block_size` is clamped to at
/// least 1 (a no-op copy at 1). Edge cells are clipped to the region.
///
/// `out` must be at least `source.width * source.height * 4` bytes. Safe to
/// pass `source.surface.data` as `out` for a full-surface region — each cell
/// is fully read before it is written.
pub fn apply_surface_pixelate_filter(out: &mut [u8], source: &SurfaceRegion, block_size: u32) {
    let block = block_size.max(1);
    let w = source.width;
    let h = source.height;
    let surface_width = source.surface.width;
    let surface_height = source.surface.height;
    let data = &source.surface.data;

    let mut by = 0;
    while by < h {
        let y_end = (by + block).min(h);
        let mut bx = 0;
        while bx < w {
            let x_end = (bx + block).min(w);
            let mut r = 0_u64;
            let mut g = 0_u64;
            let mut b = 0_u64;
            let mut a = 0_u64;
            let mut count = 0_u64;
            for py in by..y_end {
                let sy = source.y + py;
                if sy >= surface_height {
                    continue;
                }
                for px in bx..x_end {
                    let sx = source.x + px;
                    if sx >= surface_width {
                        continue;
                    }
                    let si = ((sy * surface_width + sx) * 4) as usize;
                    r += data[si] as u64;
                    g += data[si + 1] as u64;
                    b += data[si + 2] as u64;
                    a += data[si + 3] as u64;
                    count += 1;
                }
            }
            if count != 0 {
                let ar = div_round(r, count);
                let ag = div_round(g, count);
                let ab = div_round(b, count);
                let aa = div_round(a, count);
                for py in by..y_end {
                    for px in bx..x_end {
                        let di = ((py * w + px) * 4) as usize;
                        out[di] = ar;
                        out[di + 1] = ag;
                        out[di + 2] = ab;
                        out[di + 3] = aa;
                    }
                }
            }
            bx += block;
        }
        by += block;
    }
}

fn div_round(numerator: u64, denominator: u64) -> u8 {
    (numerator as f64 / denominator as f64).round() as u8
}

#[cfg(test)]
mod tests {
    use super::*;
    use flighthq_surface::create_surface;
    use flighthq_types::SurfaceRegion;

    fn region(surface: flighthq_types::Surface) -> SurfaceRegion {
        let width = surface.width;
        let height = surface.height;
        SurfaceRegion {
            surface,
            x: 0,
            y: 0,
            width,
            height,
        }
    }

    #[test]
    fn apply_surface_pixelate_filter_block_one_is_copy() {
        let mut source = create_surface(2, 1, 0);
        source.data[0] = 10;
        source.data[4] = 200;
        let mut out = vec![0_u8; 8];
        apply_surface_pixelate_filter(&mut out, &region(source), 1);
        assert_eq!(out[0], 10);
        assert_eq!(out[4], 200);
    }

    #[test]
    fn apply_surface_pixelate_filter_averages_block() {
        let mut source = create_surface(2, 1, 0);
        source.data[0] = 0;
        source.data[4] = 100;
        source.data[3] = 255;
        source.data[7] = 255;
        let mut out = vec![0_u8; 8];
        apply_surface_pixelate_filter(&mut out, &region(source), 2);
        assert_eq!(out[0], 50);
        assert_eq!(out[4], 50);
    }
}
